### ISSUE TITLE
Handle app launches from push notifications

### ISSUE THREAD

**Issue by ZachOrr** [the-blue-alliance/the-blue-alliance-ios#178]
Tracking issue for handling all server-defined push notification types in the iOS app. Initial framework landed in #1096 — the rest of the types still need their own tap actions wired through `PushNotificationRouter`.

| # | Server key | Payload keys | Proposed iOS tap action | Status |
|---|---|---|---|---|
| 1 | `upcoming_match` | event_key, match_key, team_key? | Present `MatchViewController` modally | ✅ |
| 2 | `match_score` | event_key, match_key, team_key? | Present `MatchViewController` modally | ✅ |
| 3 | `match_video` | event_key, match_key, team_key? | Present `MatchViewController` modally | ✅ |
| 4 | `starting_comp_level` | event_key, comp_level | Push `EventViewController` | ☐ |
| 5 | `alliance_selection` | event_key, team_key? | If team_key → push `TeamAtEventViewController`; else push `EventViewController` | ☐ |
| 6 | `awards_posted` | event_key, team_key? | Same as `alliance_selection` | ☐ |
| 7 | `schedule_updated` | event_key | Push `EventViewController` | ☐ |
| 8 | `district_points_updated` | district_key | Push `DistrictViewController` | ☐ |
| 9 | `event_match_video` | event_key | Push `EventViewController` | ☐ |
| 10 | `ping` | (none) | No-op on tap; just let it display | ☐ |
| 11 | `broadcast` | title, desc, url? | If `url` present, open via `urlOpener`; else no-op | ☐ |
| 12 | `media_posted` | (defined, never sent) | No-op, log it | ☐ |
| 13 | `final_results` | (defined, never sent) | No-op, log it | ☐ |
| 14 | `update_favorites` | (silent) | Background refresh of favorites store; suppress banner | ✅ |
| 15 | `update_subscriptions` | (silent) | Background refresh of subscriptions store; suppress banner | ✅ |

Server-side reference: [`src/backend/common/consts/notification_type.py`](https://github.com/the-blue-alliance/the-blue-alliance/blob/main/src/backend/common/consts/notification_type.py) and the per-type model classes under [`src/backend/common/models/notifications/`](https://github.com/the-blue-alliance/the-blue-alliance/tree/main/src/backend/common/models/notifications).

**Comment by richiksc** [comment]
This issue is still not resolved. If I get a match result notification, tapping it should go to the match details screen with score, video and breakdown. Instead, it launches the myTBA homepage.

**Comment by ZachOrr** [comment]
Yep! Correct. This functionality has not been implemented yet. It requires reworking how we instantiate the view controllers in order to require less information - since we only have a limited amount of information coming from the push notification payload.

If you’d like to take a pass at this - feel free! I’m willing to offer guidance.

**Comment by richiksc** [comment]
I'd be interested in working on this issue - but I'd need some guidance for sure! Where in the code are launches from notifications handled (I'm guessing `PushService`) and what about the `MatchViewController` would need to be changed?

**Comment by ZachOrr** [comment]
The [`PushService`](https://github.com/the-blue-alliance/the-blue-alliance-ios/blob/5a5a60bf6b6f11ec72401e17172d2b682a34c793/the-blue-alliance-ios/Services/PushService.swift#L103-L112) is setup to handle remote notifications right now - but it does not yet dispatch anywhere.

The big pieces that would need to change here is that the view controllers in TBA for iOS take full Core Data objects when instantiating (see the [`MatchViewController`](https://github.com/the-blue-alliance/the-blue-alliance-ios/blob/5a5a60bf6b6f11ec72401e17172d2b682a34c793/the-blue-alliance-ios/View%20Controllers/Match/MatchViewController.swift#L29)). The push notification does not necessarily have a full object that can be inserted - it might only contain a key for a piece of information we need. To complicate things - sometimes the push notification DOES have enough information to insert/create a Core Data model from the notification's data payload.

The recommendation here is that the view controllers should be flexible with their instantiation and take either a full Core Data object *or* a key (possibly several keys, I suppose) that would be enough information to kickoff an initial refetch for the view in order to populate the data in the views. In the bast case scenario, we would do this refresh *once* in one of the child view controllers (for parent view controllers that) and watch for the object in the other child view controllers to make sure we're doing as few network requests as possible.

Finally - the `PushService` is going to need to DI the `rootControllerProvider` so we can have some view hierarchy to show a view on top of. It's worth considering if we should push or show a modal in these situations.

**Comment by ZachOrr** [comment]
Anything labeled [`good first issue`](https://github.com/the-blue-alliance/the-blue-alliance-ios/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22) would be a good place to start
